sdks/python/apache_beam/runners/direct/direct_runner.py

```diff
@@ -516,7 +516,8 @@ def visit_transform(self, applied_ptransform):
         if isinstance(applied_ptransform.transform, ExternalTransform):
           raise RuntimeError(
               "Streaming Python direct runner "
-              "does not support cross-language pipelines.")
+              "does not support cross-language pipelines."
+              "Please use other runners such as DataflowRunner.")
```

Comment:
Routing to dataflow from an attempted use of the direct runner is a huge jump in terms of cost and commitment. This should probably direct users to the python portable runner + documentation on how to run xlang pipelines that way so we're staying local.

Reply:
@lostluck is it reasonable to recommend prism at this point? If so, is there a good doc to link to?

Reply:
The Beam codebase shouldn't be recommending a non-free runner. It's a conflict of interest WRT the governance of the project.

I would love for *Python* users to try Prism out before we make the recommendation though. All I do is run the test suites. I don't try running proper pipelines at this stage.

WRT Prism, by 2.60.0 it's hopefully passing all the Python Validates Runner tests, and all the Java Validates Runner Tests that succeed against the Python Portable Runner. Unfortunately, it's not quite the full set for 2.59.0. For 2.59.0, Prism is presently launching it as "experimental support" in the Java and Python SDKs.

Unfortunately, because I'm on the release, and I need to finish my talks for Beam Summit, making Prism documentation on the Beam site and similar isn't happening until after Beam Summit, in September.

WRT to the specific reason for this message: Prism doesn't care if something is streaming or not. It has unified execution of bounded and unbounded pcollections. So xlang is an orthogonal concern.

However, per the above, prism doesn't support *everything* yet, so triggers or specific windowing strategy configurations, among other things, may prevent pipeline execution at this time. But TBH it's probably fine for most pipelines. No "real time" processing time timers either (which also extends to ProcessContinuation delays for streaming IOs too).

Reply:
If this were to say "...such as FlinkRunner, DataflowRunner, or PrismRunner", that would avoid the conflict problem though.  (IIUC, Python will download and launch a Flink instance, just as it would for Prism).

Reply:
> I would love for Python users to try Prism out before we make the recommendation though. All I do is run the test suites. I don't try running proper pipelines at this stage.

IMO this is actually the perfect place to get users to try prism. What they've just tried didn't work, so at worst we're providing an equal experience (and hopefully we're providing a better one).

> If this were to say "...such as FlinkRunner, DataflowRunner, or PrismRunner", that would avoid the conflict problem though. (IIUC, Python will download and launch a Flink instance, just as it would for Prism).

I think this is a reasonable option, though I'd still vote we just use this as an opportunity to guide towards prism. I think that is the best way to start exposing gaps that may exist, and this seems like a low risk environment to do so

Reply:
SGTM

Reply:
SGTM. Added other runners.